tests/matrix_free/get_cell_catergory_01.cc

```diff
@@ -0,0 +1,238 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+
+// tests MatrixFree::get_cell_category() and
+// MatrixFree::get_cell_range_category()
+
+#include <deal.II/distributed/shared_tria.h>
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_nothing.h>
+#include <deal.II/fe/fe_q.h>
+
+#include <deal.II/grid/filtered_iterator.h>
+#include <deal.II/grid/grid_generator.h>
+
+#include <deal.II/matrix_free/fe_evaluation.h>
+#include <deal.II/matrix_free/matrix_free.h>
+
+#include "../tests.h"
+
+template <int dim>
+void
+test()
+{
+  using Number              = double;
+  using VectorizedArrayType = VectorizedArray<Number>;
+
+  Triangulation<dim> tria;
+
+  GridGenerator::hyper_cube(tria);
+  tria.refine_global(3);
+
+  // caterorization - not strict
+  {
+    DoFHandler<dim> dof_handler(tria);
+    dof_handler.distribute_dofs(FE_Q<dim>(1));
+
+    typename MatrixFree<dim, Number, VectorizedArrayType>::AdditionalData data;
+    data.tasks_parallel_scheme = MatrixFree<dim, double>::AdditionalData::none;
+    data.mapping_update_flags  = update_quadrature_points;
+
+    std::vector<unsigned int> cell_vectorization_category(
+      tria.n_active_cells());
+    for (unsigned int i = 0; i < cell_vectorization_category.size(); ++i)
+      cell_vectorization_category[i] = i / 7;
+
+    data.cell_vectorization_category          = cell_vectorization_category;
+    data.cell_vectorization_categories_strict = false;
+
+    MatrixFree<dim, Number, VectorizedArrayType> matrix_free;
+    matrix_free.reinit(MappingQ1<dim>(),
+                       dof_handler,
+                       AffineConstraints<Number>(),
+                       QGauss<dim>(2),
+                       data);
+
+    double dummy = 0;
+
+    matrix_free.template cell_loop<double, double>(
+      [&](const auto &matrix_free, auto &, const auto &, const auto range) {
+        unsigned int max_category_range = 0;
+
+        for (unsigned int cell = range.first; cell < range.second; ++cell)
+          {
+            unsigned int max_category = 0;
+            for (unsigned int v = 0;
+                 v < matrix_free.n_active_entries_per_cell_batch(cell);
+                 ++v)
+              {
+                const auto category = cell_vectorization_category
+                  [matrix_free.get_cell_iterator(cell, v)->active_cell_index()];
+                std::cout << category << " ";
```

Comment:
Is this output on purpose? It won't be compared, and one will only see it when manually running the test (which might be ok).

Reply:
I would keep it this way. I introduced these for visualization purposes. It looks like this:
```
0 0 0 0 0 0 0 1 
1 1 1 1 1 1 2 2 
2 2 2 2 2 3 3 3 
3 3 3 3 4 4 4 4 
4 4 4 5 5 5 5 5 
5 5 6 6 6 6 6 6 
6 7 7 7 7 7 7 7 
8 8 8 8 8 8 8 9 

0 0 0 0 0 0 0 
1 1 1 1 1 1 1 
2 2 2 2 2 2 2 
3 3 3 3 3 3 3 
4 4 4 4 4 4 4 
5 5 5 5 5 5 5 
6 6 6 6 6 6 6 
7 7 7 7 7 7 7 
8 8 8 8 8 8 8 
9 

0 0 0 0 0 0 0 

1 1 1 1 1 1 1 

2 2 2 2 2 2 2 

3 3 3 3 3 3 3 

4 4 4 4 4 4 4 

5 5 5 5 5 5 5 

6 6 6 6 6 6 6 

7 7 7 7 7 7 7 

8 8 8 8 8 8 8 

9 
```

In the test, I am only checking the properties, since outputting such an output would require that we create an output for each vectorization type.

Reply:
I agree that we should not print this with vectorization. Maybe add a short comment in the code that explains why we have the `cout` in the test, to guide the reader of test on how to use it?